PWN-V1.3 | Ethereum | 24.02.2025 (#102)
deficollective/defiscan: src/content/protocols/pwn/ethereum.md

Furthermore, users can utilise funds from other protocols with PWN Pool Hooks. All of these are on opt-in basis and their failure doesn't affect the core protocol in any way.

> Autonomy score: Low

## Exit Window

Permissions are controlled by the _PWN DAO_ and parameter changes do not affect existing running loans, therefore there is no need for an exit window. Once a user starts using the protocol (creates a loan), the rules cannot change for them. 

> Exit Window score: Low

## Accessibility

PWN can be used on the PWN Platform (app.pwn.xyz). Once a loan is created, interactions with the protocol are simple enough to be feasibly done through etherscan in case of the main interface failure. PWN additionally provides a guide for funds recovery using Etherscan: https://docs.pwn.xyz/Guides/other/recover.

> Accessibility score: Medium

## Conclusion

The PWN protocol achieves _Low_ centralization risk scores for its _Upgradeability_, _Autonomy_, _Exit Window_ and _Medium_ for the _Accessibility_ dimension. It thus ranks _Stage 1_.

The protocol could reach Stage 2 if it would have multiple independent interfaces.

# Reviewer's Notes

Nothing to note.

# Protocol Analysis

> ![Overview of the PWN protocol](../diagrams/pwn-overview.png)

Author: thanks, fixed.

| PWNHub | setTags | Same as setTag. This function calls setTag multiple times. | TimelockController (Protocol) |
| PWNLOAN | mint | Mints a LOAN token. LOAN token represents a loan in the protocol for the lender. It is required to claim repayment or collateral. Malicious actor could mint worthless LOAN tokens and try to use them as a collateral for loans. | Loan contract specified in an offer |
| PWNLOAN | burn | Burns a LOAN token.LOAN token represents a loan in the protocol for the lender. It is required to claim repayment or collateral. Malicious actor could burn other LOAN tokens and make it impossible to claim loans. | Loan contract specified in an offer |
| PWNRevokedNonce | revokeNonce | Revokes nonce for an offer. Used to revoke an offchain signed offer on behalf of an owner. Malicious actor could revoke offers that the user wants to be valid. | Contract with a valid tag in PWNHub |
| PWNRevokedNonce | revokeNonce | Revokes nonce for an offer. Used to revoke an offchain signed offer on behalf of an owner. Malicious actor could revoke offers that the user wants to be valid. | Contract with a valid tag in PWNHub |

> | PWNSimpleLoan | createLOAN | Creates a loan in the PWN Protocol. The check in this function is for a valid lenderSpec check in case the caller is a lender. There's also a reentrancy protection modifier. | Lender |
> | PWNSimpleLoan | repayLOAN | Repays a running loan. There's a reentrancy protection modifier to make sure you can't repay more than once. | Borrower |
> | PWNSimpleLoan | claimLOAN | Claims repayment or defaulted collateral. There's a reentrancy protection modifier to make sure you can't claim more than once. | Lender |
> | PWNSimpleLoan | tryClaimRepaidLOAN | Tries to claim loan for a lender. The check is this function is to make sure this function can only be called by the Loan contract. Malicious actor could initiate the claim without the Loan contract. Note that it's not possible to claim through this function if the original loan lender has changed and therefor malicious actor cannot claim the assets to its wallets. | Loan contract |
> | PWNSimpleLoan | makeExtensionProposal | Makes an onchain proposal for loan extension. The borrower can create a proposal for extension. Malicious actor could create an extension proposal for any loan. | Proposer |
> | PWNSimpleLoan | extendLOAN | Extends a loan based on extension proposal. Both the lender and the borrower can call this function to extend a loan they're part of if the other party has signed the extension proposal. The checks in this function are to distinguish two possible states (lender or borrower calling the function). | Lender or borrower |

Reviewer: we dont include user actions in protocols as part of the permissions, we didnt include user actions in aave, spark and pendle e.g
```suggestion
```

> | PWNSimpleLoanSimpleProposal | _makeProposal | Marks a proposalHash as valid. Caller has to also be a proposer of the corresponding offer. | Proposer |
> | PWNSimpleLoanSimpleProposal | _acceptProposal | Accepts a given proposal. Loan contracts call this function with the signed proposal on behalf of the acceptor. Malicious actor could accept proposals for different Loan contracts. | Loan contract specified in an offer |
> | PWNSimpleLoanSimpleProposal | makeProposal | Makes an onchain proposal. Checks in this function are derived from _makeProposal. | Proposer |
> | PWNSimpleLoanSimpleProposal | acceptProposal | Accepts a given proposal. Checks in this function are derived from _acceptProposal. | Loan contract specified in an offer |
> | PWNSimpleLoanListProposal | _makeProposal | Marks a proposalHash as valid. Caller has to also be a proposer of the corresponding offer. | Proposer |
> | PWNSimpleLoanListProposal | _acceptProposal | Accepts a given proposal. Loan contracts call this function with the signed proposal on behalf of the acceptor. Malicious actor could accept proposals for different Loan contracts. | Loan contract specified in an offer |
> | PWNSimpleLoanListProposal | makeProposal | Makes an onchain proposal. Checks in this function are derived from _makeProposal. | Proposer |
> | PWNSimpleLoanListProposal | acceptProposal | Accepts a given proposal. Checks in this function are derived from _acceptProposal. | Loan contract specified in an offer |
> | PWNSimpleLoanElasticProposal | _makeProposal | Marks a proposalHash as valid. Caller has to also be a proposer of the corresponding offer. | Proposer |
> | PWNSimpleLoanElasticProposal | _acceptProposal | Accepts a given proposal. Loan contracts call this function with the signed proposal on behalf of the acceptor. Malicious actor could accept proposals for different Loan contracts. | Loan contract specified in an offer |
> | PWNSimpleLoanElasticProposal | makeProposal | Makes an onchain proposal. Checks in this function are derived from _makeProposal. | Proposer |
> | PWNSimpleLoanElasticProposal | acceptProposal | Accepts a given proposal. Checks in this function are derived from _acceptProposal. | Loan contract specified in an offer |
> | PWNSimpleLoanDutchAuctionProposal | _makeProposal | Marks a proposalHash as valid. Caller has to also be a proposer of the corresponding offer. | Proposer |
> | PWNSimpleLoanDutchAuctionProposal | _acceptProposal | Accepts a given proposal. Loan contracts call this function with the signed proposal on behalf of the acceptor. Malicious actor could accept proposals for different Loan contracts. | Loan contract specified in an offer |
> | PWNSimpleLoanDutchAuctionProposal | makeProposal | Makes an onchain proposal. Checks in this function are derived from _makeProposal. | Proposer |
> | PWNSimpleLoanDutchAuctionProposal | acceptProposal | Accepts a given proposal. Checks in this function are derived from _acceptProposal. | Loan contract specified in an offer |
> | PWNSimpleLoanElasticChainlinkProposal | _makeProposal | Marks a proposalHash as valid. Caller has to also be a proposer of the corresponding offer. | Proposer |
> | PWNSimpleLoanElasticChainlinkProposal | _acceptProposal | Accepts a given proposal. Loan contracts call this function with the signed proposal on behalf of the acceptor. Malicious actor could accept proposals for different Loan contracts. | Loan contract specified in an offer |
> | PWNSimpleLoanElasticChainlinkProposal | makeProposal | Makes an onchain proposal. Checks in this function are derived from _makeProposal. | Proposer |
> | PWNSimpleLoanElasticChainlinkProposal | acceptProposal | Accepts a given proposal. Checks in this function are derived from _acceptProposal. | Loan contract specified in an offer |

Reviewer: ```suggestion
| PWNSimpleLoanSimpleProposal | acceptProposal | Accepts a given proposal. Loan contracts call this function with the signed proposal on behalf of the acceptor. | Loan contract specified in an offer |
| PWNSimpleLoanListProposal | acceptProposal |Accepts a given proposal. Loan contracts call this function with the signed proposal on behalf of the acceptor. | Loan contract specified in an offer |
| PWNSimpleLoanElasticProposal | acceptProposal | Accepts a given proposal. Loan contracts call this function with the signed proposal on behalf of the acceptor. | Loan contract specified in an offer |
| PWNSimpleLoanDutchAuctionProposal | acceptProposal | Accepts a given proposal. Loan contracts call this function with the signed proposal on behalf of the acceptor. | Loan contract specified in an offer |
| PWNSimpleLoanElasticChainlinkProposal | acceptProposal | Accepts a given proposal. Loan contracts call this function with the signed proposal on behalf of the acceptor. | Loan contract specified in an offer |
```

Reviewer: summarise _acceptProposal and acceptProposal, remove user only facing external/public endpoints

> | PWNUtilizedCredit | utilizeCredit | Updates utilised credit for an account. This is used to share credit between different proposal types. Malicious actor could update the utilised credit to go over the credit limit. | Contract with a valid tag in PWNHub |

Reviewer: ```suggestion
| PWNUtilizedCredit | utilizeCredit | Updates utilised credit for an account. This is used to share credit between different proposal types. Malicious actor would add a valid tag to a malicious contract which could update the utilised credit to go over the credit limit. | Contract with a valid tag in PWNHub |
```

| TimelockController (Protocol) | schedule | Schedules a transaction to be executed after the mandatory delay has passed (currently none). This transaction can perform any action that this contract is allowed to, including upgrading PWNConfig and tagging new contracts in PWNHub. | PWNDAO |

> | TimelockController (Protocol) | scheduleBatch | Similar toschedule, for a batch of transactions. | PWNDAO |

Reviewer: ```suggestion
| TimelockController (Protocol) | scheduleBatch | Similar to schedule, for a batch of transactions. | PWNDAO |
```

| TimelockController (Protocol) | updateDelay | Updates the mandatory delay. There are no minimum or maximum delays enforced in the contract. | TimelockController |
| TimelockController (Protocol) | cancel | Cancels a queued transaction before it has been executed. | PWNDAO |
| TimelockController (Protocol) | grantRole | Grants a role to a given address. There are different roles to schedule, cancel, and execute transactions, currently executing transactions is open to anyone. | TimelockController (Protocol) |
| TimelockController (Protocol) | revokeRole| Revokes a role from a given address. | TimelockController (Protocol) |
| TimelockController (Admin) | schedule | Schedules a transaction to be executed after the mandatory delay has passed (currently none). This transaction can perform any action that this contract is allowed to, including changing parameters in PWNConfig. | PWNDAO |

> | TimelockController (Admin) | scheduleBatch | Similar toschedule, for a batch of transactions. | PWNDAO |

Reviewer: ```suggestion
| TimelockController (Admin) | scheduleBatch | Similar to schedule, for a batch of transactions. | PWNDAO |
```
